File: CanvasCore/CanvasCore/Helm/Helm.m
//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

#import <React/RCTBridgeModule.h>
#import <CanvasCore/CanvasCore-Swift.h>

@interface Helm: NSObject <RCTBridgeModule>

@end

@implementation Helm

RCT_EXPORT_MODULE();

- (dispatch_queue_t)methodQueue {
    return dispatch_get_main_queue();
}

RCT_EXPORT_METHOD(setScreenConfig:(NSDictionary *)config forScreenWithID:(NSString *)screenInstanceID hasRendered:(BOOL)hasRendered) {
    [[HelmManager shared] setScreenConfig:config forScreenWithID:screenInstanceID hasRendered:hasRendered];
}

RCT_EXPORT_METHOD(setDefaultScreenConfig:(NSDictionary *)config forModule:(NSString *)module) {
    [[HelmManager shared] setDefaultScreenConfig:config forModule:module];
}

RCT_REMAP_METHOD(openInSafariViewController, openInSafariViewControllerURL:(NSURL *)url resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [[HelmManager shared] openInSafariViewController:url completion:^() {
        resolve(nil);
    }];
}

RCT_REMAP_METHOD(pushFrom, pushModule:(NSString *)sourceModule destinationModule:(NSString*)module withProps:(NSDictionary *)props options:(NSDictionary *)options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [[HelmManager shared] pushFrom:sourceModule destinationModule:module withProps:props options:options callback:^() {
        resolve(nil);
    }];
}

RCT_REMAP_METHOD(popFrom, popModule:(NSString *)sourceModule resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [[HelmManager shared] popFrom:sourceModule callback:^() {
        resolve(nil);
    }];
}

RCT_REMAP_METHOD(present, presentModule:(NSString *)module withProps:(NSDictionary *)props options:(NSDictionary *)options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [[HelmManager shared] present:module withProps:props options:options callback:^() {
        resolve(nil);
    }];
}

RCT_REMAP_METHOD(dismiss, dismissWithOptions:(NSDictionary *)options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [[HelmManager shared] dismiss:options callback:^() {
        resolve(nil);
    }];
}

RCT_REMAP_METHOD(dismissAllModals, dismissAllModalsWithOptions:(NSDictionary *)options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    [[HelmManager shared] dismissAllModals:options callback:^() {
        resolve(nil);
    }];
}

RCT_EXPORT_METHOD(traitCollection:(NSString*)moduleName callback:(RCTResponseSenderBlock)callback) {
    [[HelmManager shared] traitCollection:moduleName callback:callback];
}

RCT_EXPORT_METHOD(loginComplete) {
    [[HelmManager shared] loginComplete];
}

RCT_EXPORT_METHOD(reload) {
    [[HelmManager shared] reload];
}

RCT_EXPORT_METHOD(registerRoute:(NSString *)template) {
    [[HelmManager shared] registerRoute:template];
}

@end

File: CanvasCore/CanvasCore/Helm/Helm.swift
//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import UIKit
import SafariServices
import React
import Core

public typealias ModuleName = String

public struct HelmViewControllerFactory {
    public typealias Props = [String: Any]
    public typealias Builder = (Props) -> UIViewController?
}


@objc(HelmManager)
open class HelmManager: NSObject {

    @objc public static let shared = HelmManager()
    @objc public var bridge: RCTBridge!
    @objc public var onReactLoginComplete: () -> Void = {}
    @objc public var onReactReload: () -> Void = {}

    @objc
    public func loginComplete() {
        onReactLoginComplete()
    }

    //  MARK: - Init

    override init() {
        super.init()
        setupNotifications()
    }

    @objc func setupNotifications() {
        NotificationCenter.default.addObserver(self, selector: #selector(reactWillReload), name: Notification.Name("RCTJavaScriptWillStartLoadingNotification"), object: nil)
    }

    @objc open func reload() {
        RCTTriggerReloadCommandListeners("reload called")
    }

    @objc open func reactWillReload() {}

    //  MARK: - Screen Configuration

    @objc open func registerNativeViewController(for moduleName: ModuleName, factory: @escaping HelmViewControllerFactory.Builder) {}

    func register<T: UIViewController>(screen: T) {}

    @objc open func registerRoute(_ template: String) {}

    @objc open func setScreenConfig(_ config: [String: Any], forScreenWithID screenInstanceID: String, hasRendered: Bool) {}

    @objc open func setDefaultScreenConfig(_ config: [String: Any], forModule module: ModuleName) {}

    //  MARK: - Navigation

    @objc public func openInSafariViewController(_ url : URL, completion: @escaping () -> Void) {}

    @objc public func pushFrom(_ sourceModule: ModuleName, destinationModule: ModuleName, withProps props: [String: Any], options: [String: Any], callback: (() -> Void)? = nil) {}

    @objc open func popFrom(_ sourceModule: ModuleName, callback: (() -> Void)? = nil) {}

    @objc public func present(_ module: ModuleName, withProps props: [String: Any], options: [String: Any], callback: (() -> Void)? = nil) {}

    @objc open func dismiss(_ options: [String: Any], callback: (() -> Void)? = nil) {}

    @objc open func dismissAllModals(_ options: [String: Any], callback: (() -> Void)? = nil) {}

    @objc public func traitCollection(_ moduleName: String, callback: @escaping RCTResponseSenderBlock) {}
    
    @objc public func cleanup(callback: (() -> Void)?) {}

    public func routeHandlers(_ routeMap: KeyValuePairs<String, RouteHandler.ViewFactory?>) -> [RouteHandler] { [] }
}

extension HelmManager {
    @objc open func topMostViewController() -> UIViewController? {
        return AppEnvironment.shared.window?.rootViewController?.topMostViewController()
    }
}
